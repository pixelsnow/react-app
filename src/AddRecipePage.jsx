import React, { useEffect, useState } from "react";
import axios from "axios";
import { NavLink } from "react-router-dom";
import cloneDeep from "lodash/cloneDeep";

import classes from "./AddRecipePage.module.css";

const emptyRecipe = {
  id: undefined,
  name: "",
  author: "",
  country: "",
  description: "",
  image: "",
  ingredients: [
    {
      ingredient: "",
      quantity: "",
    },
  ],
  instructions: "",
};

const AddRecipePage = () => {
  const [recipe, setRecipe] = useState(cloneDeep(emptyRecipe));
  const [recipePosted, setRecipePosted] = useState(false);
  const [id, setId] = useState();
  const [allCountries, setAllCountries] = useState([]);

  const changeHandler = (e) => {
    setRecipe({ ...recipe, [e.target.name]: e.target.value });
  };

  const changeIngredient = (e, index) => {
    const newRecipe = cloneDeep(recipe);
    newRecipe.ingredients[index][e.target.name] = e.target.value;
    setRecipe(newRecipe);
  };

  const addIngredient = (e) => {
    e.preventDefault();
    const newRecipe = cloneDeep(recipe);
    newRecipe.ingredients.push({
      ingredient: "",
      quantity: "",
    });
    setRecipe(newRecipe);
  };

  const removeIngredient = (e, index) => {
    e.preventDefault();
    if (recipe.ingredients.length <= 1) return;
    const newRecipe = cloneDeep(recipe);
    newRecipe.ingredients.splice(index, 1);
    setRecipe(newRecipe);
  };

  const handlePost = (e) => {
    e.preventDefault();
    setRecipePosted(true);
    axios
      .post("http://localhost:4000/recipes", recipe)
      .then((res) => {
        setId(res.data.id);
      })
      .catch((err) => {
        alert("Server error");
      });
  };

  useEffect(() => {
    axios
      .get("https://restcountries.com/v3.1/all")
      .then((data) => {
        let fetchedCountries = [];
        data.data.forEach((country) => {
          fetchedCountries.push(country.name.common);
        });
        fetchedCountries.sort();
        setAllCountries(fetchedCountries);
      })
      .catch((err) => alert("Country list couldn't be loaded"));
  }, [allCountries]);

  const addAnother = () => {
    setRecipePosted(false);
    setRecipe(cloneDeep(emptyRecipe));
  };

  return (
    <div className={classes.form_page}>
      {!recipePosted && (
        <div>
          <h1>Add your recipe</h1>
          <form className={classes.form} onSubmit={(e) => handlePost(e)}>
            <label htmlFor="name">Title</label>
            <input
              autoFocus
              required
              maxLength="40"
              id="name"
              name="name"
              type="text"
              onChange={(e) => changeHandler(e)}
            />
            <label htmlFor="author">Your name</label>
            <input
              maxLength="25"
              id="author"
              name="author"
              type="text"
              onChange={(e) => changeHandler(e)}
            />
            <label htmlFor="description">Description</label>
            <textarea
              maxLength="100000"
              id="description"
              name="description"
              cols="30"
              rows="7"
              onChange={(e) => changeHandler(e)}
            ></textarea>
            <label htmlFor="country">Country</label>
            <select
              id="country"
              name="country"
              defaultValue={""}
              onChange={(e) => changeHandler(e)}
            >
              <option value="">Select country</option>
              {allCountries.map((country) => (
                <option key={country} value={country}>
                  {country}
                </option>
              ))}
            </select>
            {recipe.ingredients.length > 0 && (
              <div className={classes.ingredient_labels}>
                <label
                  style={{
                    width: recipe.ingredients.length <= 1 ? "51%" : "47%",
                  }}
                  htmlFor="ingredient"
                >
                  Ingredient
                </label>
                <label htmlFor="quantity">Quantity</label>
              </div>
            )}
            {recipe.ingredients.map((ingredient, index) => (
              <div key={index} className={classes.ingredient_container}>
                <div
                  className={
                    classes.ingredient_field +
                    (recipe.ingredients.length <= 1 ? " only" : "")
                  }
                >
                  <input
                    required
                    id="ingredient"
                    name="ingredient"
                    type="text"
                    value={recipe.ingredients[index].ingredient}
                    onChange={(e) => changeIngredient(e, index)}
                  />
                </div>
                <div
                  className={
                    classes.ingredient_field +
                    (recipe.ingredients.length <= 1 ? " only" : "")
                  }
                >
                  <input
                    id="quantity"
                    name="quantity"
                    type="text"
                    value={recipe.ingredients[index].quantity}
                    onChange={(e) => changeIngredient(e, index)}
                  />
                </div>
                {recipe.ingredients.length > 1 && (
                  <button onClick={(e) => removeIngredient(e, index)}>
                    <span className="material-symbols-outlined">close</span>
                  </button>
                )}
              </div>
            ))}
            {recipe.ingredients.length < 50 && (
              <button
                className={classes.add_button}
                onClick={(e) => addIngredient(e)}
              >
                +
              </button>
            )}
            <label htmlFor="image">Image link</label>
            <input
              maxLength="3000"
              id="image"
              name="image"
              type="text"
              onChange={(e) => changeHandler(e)}
            />
            <label htmlFor="instructions">Instructions</label>
            <textarea
              required
              maxLength="100000"
              id="instructions"
              name="instructions"
              cols="30"
              rows="7"
              onChange={(e) => changeHandler(e)}
            ></textarea>
            <div className={classes.submit_container}>
              <button className={classes.large_button} type="submit">
                Submit
              </button>
            </div>
          </form>
        </div>
      )}

      {recipePosted && (
        <div className={classes.confirmation}>
          <p>Recipe {recipe.name} has been added successfully!</p>
          <div>
            <NavLink to={`/${id}`}>
              <button>See your recipe</button>
            </NavLink>
            <button onClick={addAnother}>Add another one</button>
          </div>
        </div>
      )}
    </div>
  );
};

export default AddRecipePage;
