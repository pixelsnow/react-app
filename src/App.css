@import url("https://fonts.googleapis.com/css2?family=Lora&family=PT+Sans&display=swap");

:root {
  --card-height: 420px;
  --card-width: 300px;
}

html {
  background: linear-gradient(to bottom, rgb(87, 87, 87), rgb(33, 33, 33))
    no-repeat center center fixed;
}

h1,
h2,
h3 {
  font-family: "Lora", serif;
  font-weight: 400;
}

input,
select,
textarea {
  font-family: inherit;
  font-size: inherit;
  border: none;
  padding: 0 1rem;
  border-radius: 0.3rem;
}

input,
select {
  height: 3rem;
}

button {
  font-family: inherit;
  font-size: inherit;
  border: none;
  padding: 0.5rem 1rem;
  border-radius: 0.3rem;
  background-color: black;
  color: white;
}
