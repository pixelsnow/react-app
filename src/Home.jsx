import React from "react";
import { NavLink } from "react-router-dom";

import Hero from "./Hero";

import classes from "./Home.module.css";

const Home = () => {
  return (
    <div>
      <Hero />
      <div className={classes.cta_container}>
        <NavLink to="/recipes">
          <div className={classes.cta}>
            <h3>Browse recipes</h3>
            <p>Travel the word from your kitchen</p>
          </div>
        </NavLink>
        <NavLink to="/add_recipe">
          <div className={classes.cta}>
            <h3>Add your own recipe</h3>
            <p>
              Your country's food is missing from the database? Share your
              favourite meals with us{" "}
              <span className="material-symbols-outlined">favorite</span>
            </p>
          </div>
        </NavLink>
        <a href="https://en.bc.fi/" target="_blank" rel="noreferrer">
          <div className={classes.cta}>
            <h3>Want to know more?</h3>
            <p>Visit Business College Helsinki Website</p>
          </div>
        </a>
      </div>
    </div>
  );
};

export default Home;
