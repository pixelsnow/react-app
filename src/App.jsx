import { BrowserRouter, Route, Routes } from "react-router-dom";

import Layout from "./Layout";
import Home from "./Home";
import Recipes from "./Recipes";
import RecipePage from "./RecipePage";
import About from "./About";

import "./App.css";

function App() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Layout />}>
          <Route index element={<Home />} />
          <Route path="recipes" element={<Recipes />} />
          <Route path="about" element={<About />} />
          <Route path=":categoryName" element={<RecipePage />}></Route>
        </Route>
      </Routes>
    </BrowserRouter>
  );
}

export default App;
