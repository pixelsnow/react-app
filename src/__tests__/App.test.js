import React from "react";
import { BrowserRouter } from "react-router-dom";

import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import App from "../App";
import Home from "../Home";
import AddRecipePage from "../AddRecipePage";

// Making sure that all 3 cards on the home page are rendered correctly
describe("Home component - testing that cards are rendered", () => {
  test("Exactly 3 headings of 3rd level exist", () => {
    render(<Home />, { wrapper: BrowserRouter });
    const titles = screen.getAllByRole("heading", { level: 3 });
    expect(titles.length).toBe(3);
  });
  test("Heading 1 is rendered correctly", () => {
    render(<Home />, { wrapper: BrowserRouter });
    const title1 = screen.getByText("Browse recipes");
    expect(title1).toBeInTheDocument();
  });
  test("Heading 2 is rendered correctly", () => {
    render(<Home />, { wrapper: BrowserRouter });
    const title2 = screen.getByText("Add your own recipe");
    expect(title2).toBeInTheDocument();
  });
  test("Heading 3 is rendered correctly", () => {
    render(<Home />, { wrapper: BrowserRouter });
    const title3 = screen.getByText("Want to know more?");
    expect(title3).toBeInTheDocument();
  });
});

// Snapshot test
describe("Snapshot test - App front page", () => {
  test("App matches Snapshot", () => {
    const domTree = renderer.create(<App />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});

// Navigation test - mimicking user navigating the website
describe("Navigation test", () => {
  test("Navigate to about page, rendered correctly", async () => {
    render(<App />);
    const user = userEvent.setup();
    // Verifying page content for the About page
    await user.click(screen.getByText(/About/i));
    expect(screen.getByText(/About TasteIT/i)).toBeInTheDocument();
  });

  test("Navigate to recipes page, rendered correctly", async () => {
    render(<App />);
    const user = userEvent.setup();
    // Verifying page content for the recipes page
    await user.click(screen.getByText(/Browse Recipes/i));
    expect(screen.getByText(/Search recipes/i)).toBeInTheDocument();
  });

  test("Navigate to add recipe page, rendered correctly", async () => {
    render(<App />);
    const user = userEvent.setup();
    // Verifying page content for add recipe page
    await user.click(screen.getByText(/Add a recipe/i));
    expect(screen.getByText(/Add your recipe/i)).toBeInTheDocument();
  });
});

describe("Form test", () => {
  test("Checking that form is rendered properly", () => {
    render(<AddRecipePage />);
    expect(screen.getByText(/^your name$/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/^title$/i)).toBeInTheDocument();
  });

  test("Checking that form successfully submits with correct data", async () => {
    render(<AddRecipePage />, { wrapper: BrowserRouter });
    const titleInput = screen.getByLabelText(/title/i);
    console.log(titleInput);
    const instructionsInput = screen.getByLabelText(/instructions/i);
    const submitButton = screen.getByRole("button", { name: "Submit" });
    userEvent.type(titleInput, "New recipe");
    userEvent.type(instructionsInput, "Cook and eat!");
    await userEvent.click(submitButton);
    const confirmation = screen.getByText("See your recipe");
    expect(confirmation).toBeInTheDocument();
  });
});

// simulating user interaction
/*
test("allows users to add items to their list", () => {
  const { getByText, getByLabelText } = render(<App />);

  const input = getByLabelText("What needs to be done?");
  const button = getByText("Add #1");

  fireEvent.change(input, { target: { value: "Feed cats" } });
  fireEvent.click(button);
  // look at @testing-library/user-event for more interactions
  // https://testing-library.com/docs/user-event/intro

  getByText("Feed cats");
  getByText("Add #2");
});
*/

test("Opens BCH link on a new tab", () => {
  render(<App />);
  //const collegeLink = screen.getByText("");
  // expect(collegeLink.getAttribute("href").toBe("http://bc.fi"));
  /* document.addEventListener("DOMContentLoaded", function() {
      expect(document.querySelector('a').getAttribute('href')).toBe('http://bc.fi');
    }); */
});
