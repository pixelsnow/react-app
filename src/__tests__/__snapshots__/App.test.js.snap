// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test - App front page App matches Snapshot 1`] = `
<div
  className="layout"
>
  <header
    className="header"
  >
    <a
      aria-current="page"
      className="logo_link active"
      href="/"
      onClick={[Function]}
    >
      <div
        className="logo_container"
      >
        <span
          className="material-symbols-outlined"
        >
          local_dining
        </span>
        <h2>
          TasteIT
        </h2>
      </div>
    </a>
    <nav
      className="nav"
    >
      <ul>
        <a
          className=""
          href="/recipes"
          onClick={[Function]}
        >
          <li>
            Browse recipes
          </li>
        </a>
        <a
          className=""
          href="/add_recipe"
          onClick={[Function]}
        >
          <li>
            Add a recipe
          </li>
        </a>
        <a
          className=""
          href="/about"
          onClick={[Function]}
        >
          <li>
            About
          </li>
        </a>
      </ul>
    </nav>
  </header>
  <main
    className="main"
  >
    <div>
      <div
        className="hero"
      >
        <video
          autoPlay={true}
          className="bg_video"
          loop={true}
          muted={true}
        >
          <source
            src="pot.mp4"
            type="video/mp4"
          />
        </video>
      </div>
      <div
        className="cta_container"
      >
        <a
          className=""
          href="/recipes"
          onClick={[Function]}
        >
          <div
            className="cta"
          >
            <h3>
              Browse recipes
            </h3>
            <p>
              Travel the word from your kitchen
            </p>
          </div>
        </a>
        <a
          className=""
          href="/add_recipe"
          onClick={[Function]}
        >
          <div
            className="cta"
          >
            <h3>
              Add your own recipe
            </h3>
            <p>
              Your country's food is missing from the database? Share your favourite meals with us
               
              <span
                className="material-symbols-outlined"
              >
                favorite
              </span>
            </p>
          </div>
        </a>
        <a
          href="https://en.bc.fi/"
          rel="noreferrer"
          target="_blank"
        >
          <div
            className="cta"
          >
            <h3>
              Want to know more?
            </h3>
            <p>
              Visit Business College Helsinki Website
            </p>
          </div>
        </a>
      </div>
    </div>
  </main>
  <footer
    className="footer"
  >
    by Valeria Vagapova
  </footer>
</div>
`;
